### PR TITLE
Allow parsers initialize asynchronously

### PR DESCRIPTION
**Pull request by fisker**
## Description

To ship ESM version for Node.js, the original `parsers` use dynamic require to load plugins, we have to change it to `import()`.
I don't think it's a good idea to return Promise in getters. 
My initial idea is to add `hasParser()` (sync), `getParser()` (async) to plugin, then I think make them return a function  should also work. Not sure what's the best, I'd like to hear your opinions. @thorn0 @sosukesuzuki 

~I only changed JS parsers, also~ They still importing UMD version, let's make it step by step.

## Checklist

- [ ] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [ ] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by fisker** [comment]
I need fix the test, and update how to get builtin parsers after we decide what the `.parsers` should look like.

**Comment by thorn0** [comment]
Using functions makes sense.

**Review by thorn0 — APPROVED** [review]

File: src/utils/create-parsers.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+function createParsers(modules) {
```
Comment (sosukesuzuki) [inline src/utils/create-parsers.js:1]: Can you add type definition for this function?

```
/**
 * @param {Array<[() => Promise<any>, string[]]>} modules
 * @returns {{[key: string]: () => Promise<any>}}
 */
```

**Comment by sosukesuzuki** [comment]
> They still importing UMD version

Sorry, what do you mean?

**Comment by fisker** [comment]
> > They still importing UMD version
> 
> Sorry, what do you mean?

We can import esm version after we merge this

```diff
-import("./parser-graphql.js")
+import("./esm/parser-graphql.mjs")
```

File: src/utils/create-parsers.js (hunk 1 above)
Comment (fisker) [reply src/utils/create-parsers.js:1]: I don't think this is correct, we know modules[0][0] is function returns `import()`, so it's a promise resolves to a "module" with "default" and the default export contains `parsers`. The return type is also known. But I don't know how to do it.

**Comment by fisker** [comment]
I'm going to merge this, so I can continue work on migrate the bundle to ESM.

@sosukesuzuki If you want add types for `createParsers`, you can send PR https://github.com/prettier/prettier/pull/13449#discussion_r971452653

**Pull request merged**